src/install/dracut-install.c

```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
         return true;
 }
 
-static int install_dependent_module(struct kmod_module *mod, int *err)
+static int find_module_from_sysfs_node(struct kmod_ctx *ctx, const char *sysfs_node, int sysfs_node_len,
+                                       struct kmod_list **modules)
+{
+        char modalias_path[PATH_MAX] = {0};
+        snprintf(modalias_path, sizeof(modalias_path), "%.*s/modalias", sysfs_node_len, sysfs_node);
+
+        _cleanup_close_ int modalias_file = -1;
+        if ((modalias_file = open(modalias_path, O_RDONLY | O_CLOEXEC)) == -1)
+                return 0;
+
+        char alias[getpagesize()];
+        ssize_t len = read(modalias_file, alias, sizeof(alias));
+        alias[len - 1] = '\0';
+
+        return kmod_module_new_from_lookup(ctx, alias, modules);
+}
+
+static int find_suppliers_for_sys_node(struct kmod_ctx *ctx, Hashmap *suppliers, const char *node_path_raw,
+                                       size_t node_path_len)
+{
+        char node_path[PATH_MAX] = {};
+        memcpy(node_path, node_path_raw, node_path_len);
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
```

Comment:
PATH_MAX is quite big to throw it around like this, you use the same buffer for node_path and supplier_path here.